python/dolfinx/mesh.py

```diff
@@ -67,10 +66,140 @@
 ]
 
 
+class Topology:
+    """Topology for a :class:`dolfinx.mesh.Mesh`"""
+
+    _cpp_object: _cpp.mesh.Topology
+
+    def __init__(self, topology: _cpp.mesh.Topology):
+        """Initialize a topology from a C++ topology.
+        Args: The C++ topology object
+        Note:
+            topology objects should not usually be created using this
+            initializer directly.
+        """
+        self._cpp_object = topology
+
+    def cell_name(self) -> str:
+        """String representation of the cell-type of the topology"""
+        return self._cpp_object.cell_name()
+
+    def connectivity(self, d0: int, d1: int) -> _cpp.graph.AdjacencyList_int32:
+        """Return connectivity from entities of dimension ``d0`` to entities of dimension ``d1``.
+
+        Note:
+            Assumes only one entity type per dimension.
+
+        Args:
+            d0: Dimension of entity one is mapping from
+            d1: Dimension of entity one is mapping to
+        """
+        if (conn := self._cpp_object.connectivity(d0, d1)) is not None:
+            return conn
+        else:
+            raise RuntimeError(
+                f"Connectiivty between dimension {d0} and {d1} has not been computed."
+            )
+
+    @property
+    def comm(self):
+        return self._cpp_object.comm
+
+    def create_connectivity(self, d0: int, d1: int):
+        """Create connectivity between given pair of dimensions, ``d0`` and  ``d1``.
+
+        Args:
+            d0: Dimension of entity one is mapping from
+            d1: Dimension of entity one is mapping to
+        """
+        self._cpp_object.create_connectivity(d0, d1)
+
+    def create_entities(self, dim: int) -> int:
+        """Create entities of given topological dimension.
+
+        Args:
+            dim: Topological dimension
+
+        Returns:
+            Number of newly created entities, returns -1 if entities already existed
+        """
+        return self._cpp_object.create_entities(dim)
+
+    def create_entity_permutations(self):
+        """Compute entity permutations and reflections."""
+        self._cpp_object.create_entity_permutations()
+
+    @property
+    def dim(self) -> int:
+        """Return the topological dimension of the mesh."""
+        return self._cpp_object.dim
+
+    @property
+    def entity_types(self) -> list[list[CellType]]:
+        """Get the entity types in the topology for all topological dimensions."""
+        return self._cpp_object.entity_types
+
+    def get_cell_permutation_info(self) -> npt.NDArray[np.uint32]:
```

Comment:
The content of each np.uint32 being interpreted as a bitfield should probably be mentioned.

I can only find https://numpy.org/doc/stable/reference/generated/numpy.unpackbits.html in numpy which is for uint8 only.

Reply:
Yes, there doesn't seem to be a way of unpacking nicely:
https://stackoverflow.com/questions/18296035/how-to-extract-the-bits-of-larger-numeric-numpy-data-types